modules/rostests/apitests/ntdll/NtStartProfile.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * PROJECT:     ReactOS API Tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Test for NtStartProfile
+ * COPYRIGHT:   Copyright 2021 Thomas Faber (thomas.faber@reactos.org)
+ */
+
+#include "precomp.h"
+#include <versionhelpers.h>
+
+#define SIZEOF_MDL (5 * sizeof(PVOID) + 2 * sizeof(ULONG))
+typedef ULONG_PTR PFN_NUMBER;
+/* Maximum size that can be described by an MDL on 2003 and earlier */
+#define MAX_MDL_BUFFER_SIZE ((MAXUSHORT - SIZEOF_MDL) / sizeof(PFN_NUMBER) * PAGE_SIZE + PAGE_SIZE - 1)
+
+static BOOL IsWow64;
+static KAFFINITY SystemAffinityMask;
+static ULONG DummyBuffer[4096];
+
+/* The "Buffer[Offset]++;" should likely be within 128 bytes of the start
+ * of the function on any architecture we support
+ */
+#define LOOP_FUNCTION_SIZE_SHIFT 7
+#define LOOP_FUNCTION_SIZE (1UL << LOOP_FUNCTION_SIZE_SHIFT)
+C_ASSERT(LOOP_FUNCTION_SIZE == 128);
+typedef void LOOP_FUNCTION(volatile ULONG *, ULONG, ULONG);
+static
+void
+LoopFunction(
+    _Inout_updates_all_(BufferSize) volatile ULONG *Buffer,
+    _In_ ULONG BufferSizeInElements,
+    _In_ ULONG LoopCount)
+{
+    ULONG i;
+    ULONG Offset;
+
+    for (i = 0; i < LoopCount; i++)
+    {
+        for (Offset = 0; Offset < BufferSizeInElements; Offset++)
+        {
+            Buffer[Offset]++;
+        }
+    }
+}
+
+static
+void
+ProfileLoopFunction(
+    _In_ LOOP_FUNCTION *Function,
+    _Out_writes_bytes_(BufferSize) PULONG Buffer,
+    _In_ ULONG BufferSize,
+    _In_range_(0, BufferSize / sizeof(ULONG)) ULONG BufferOffset
+    )
+{
+    NTSTATUS Status;
+    HANDLE ProfileHandle;
+    ULONG Buffer1Value;
+
+    Status = NtCreateProfile(&ProfileHandle,
+                             NtCurrentProcess(),
+                             (PVOID)((ULONG_PTR)Function - LOOP_FUNCTION_SIZE),
```

Comment:
If we started at the beginning of the function, the test would not be showing anything about the bucket size. By putting the actual loop in the middle of the buffer, we can inspect the buffer values before/after and show that the bucketizing works as expected.
(The test doesn't verify the elements other than `BufferOffset` because we know nothing about how the compiler lays out the code, so we _could_ get samples hitting before/after. A future addition may address this by profiling pieces of code written in assembly, where we can create a guarantee -- but it seems useful to have this portable test as a first step)